File: packages/rslint-test-tools/tests/typescript-eslint/rules/no-unsafe-function-type.test.ts
import { RuleTester, getFixturesRootDir } from '../RuleTester.ts';

const rootPath = getFixturesRootDir();

const ruleTester = new RuleTester({
  languageOptions: {
    parserOptions: {
      project: './tsconfig.json',
      tsconfigRootDir: rootPath,
    },
  },
});

ruleTester.run('no-unsafe-function-type', {
  valid: [
    'let value: () => void;',
    'let value: <T>(t: T) => T;',
    `
      // create a scope since it's illegal to declare a duplicate identifier
      // 'Function' in the global script scope.
      {
        type Function = () => void;
        let value: Function;
      }
    `,
  ],
  invalid: [
    {
      code: 'let value: Function;',
      errors: [
        {
          column: 12,
          line: 1,
          messageId: 'bannedFunctionType',
        },
      ],
    },
    {
      code: 'let value: Function[];',
      errors: [
        {
          column: 12,
          line: 1,
          messageId: 'bannedFunctionType',
        },
      ],
    },
    {
      code: 'let value: Function | number;',
      errors: [
        {
          column: 12,
          line: 1,
          messageId: 'bannedFunctionType',
        },
      ],
    },
    {
      code: `
        class Weird implements Function {
          // ...
        }
      `,
      errors: [
        {
          column: 32,
          line: 2,
          messageId: 'bannedFunctionType',
        },
      ],
    },
    {
      code: `
        interface Weird extends Function {
          // ...
        }
      `,
      errors: [
        {
          column: 33,
          line: 2,
          messageId: 'bannedFunctionType',
        },
      ],
    },
  ],
});

File: internal/rules/no_unsafe_function_type/no_unsafe_function_type_test.go
package no_unsafe_function_type

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/rule_tester"
	"github.com/web-infra-dev/rslint/internal/rules/fixtures"
)

func TestNoUnsafeFunctionType(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &NoUnsafeFunctionTypeRule, []rule_tester.ValidTestCase{
		{
			Code: `let value: () => void;`,
		},
		{
			Code: `let value: <T>(t: T) => T;`,
		},
		{
			Code: `
// create a scope since it's illegal to declare a duplicate identifier
// 'Function' in the global script scope.
{
  type Function = () => void;
  let value: Function;
}`,
		},
	}, []rule_tester.InvalidTestCase{
		{
			Code: `let value: Function;`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "bannedFunctionType",
					Line:      1,
					Column:    12,
				},
			},
		},
		{
			Code: `let value: Function[];`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "bannedFunctionType",
					Line:      1,
					Column:    12,
				},
			},
		},
		{
			Code: `let value: Function | number;`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "bannedFunctionType",
					Line:      1,
					Column:    12,
				},
			},
		},
		{
			Code: `
class Weird implements Function {
  // ...
}`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "bannedFunctionType",
					Line:      2,
					Column:    24,
				},
			},
		},
		{
			Code: `
interface Weird extends Function {
  // ...
}`,
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "bannedFunctionType",
					Line:      2,
					Column:    25,
				},
			},
		},
	})
}
